PIC fixes for MySQL. Makes the ASM PIC-compatible, so that it works with PaX.
See http://bugs.gentoo.org/show_bug.cgi?id=42968 This patch authored by
pageexec AT freemail.hu. Ported to 4.1.12 by robbat2@gentoo.org.

diff -NuarwbB mysql.orig/strings/longlong2str-x86.s mysql/strings/longlong2str-x86.s
--- mysql.orig/strings/longlong2str-x86.s	2005-05-13 04:32:11.000000000 -0700
+++ mysql/strings/longlong2str-x86.s	2005-05-17 00:13:25.000000000 -0700
@@ -19,6 +19,13 @@
 	.file	"longlong2str.s"
 	.version "1.01"
 
+	.section	rodata
+	.align 32
+	.type	_dig_vec, @object
+	.size	_dig_vec, 37
+_dig_vec:
+	.string "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
 .text
 	.align 4
 
@@ -31,11 +38,14 @@
 	pushl %esi
 	pushl %edi
 	pushl %ebx
+
+	call __i686.get_pc_thunk.bx
+	addl $_GLOBAL_OFFSET_TABLE_,%ebx
+
 	movl 100(%esp),%esi	# Lower part of val 
 	movl 104(%esp),%ebp	# Higher part of val 
 	movl 108(%esp),%edi	# get dst 
-	movl 112(%esp),%ebx	# Radix 
-	movl %ebx,%eax
+	movl 112(%esp),%eax	# Radix 
 	testl %eax,%eax
 	jge .L144
 
@@ -50,7 +60,7 @@
 	adcl $0,%ebp
 	negl %ebp
 .L146:
-	negl %ebx		# Change radix to positive 
+	negl 112(%esp)		# Change radix to positive 
 	jmp .L148
 	.align 4
 .L144:
@@ -77,13 +87,13 @@
 
 	movl %ebp,%eax		# High part of value 
 	xorl %edx,%edx
-	divl %ebx
+	divl 112(%esp)
 	movl %eax,%ebp
 	movl %esi,%eax
-	divl %ebx
+	divl 112(%esp)
 	decl %ecx
 	movl %eax,%esi		# quotent in ebp:esi 
-	movb _dig_vec_upper(%edx),%al   # al is faster than dl 
+	movb _dig_vec@GOTOFF(%ebx,%edx),%al   # al is faster than dl 
 	movb %al,(%ecx)		# store value in buff 
 	.align 4
 .L155:
@@ -93,14 +103,13 @@
 	jl .L153
 	je .L10_mov		# Ready 
 	movl %esi,%eax
-	movl $_dig_vec_upper,%ebp
 	.align 4
 
 .L154:				# Do rest with integer precision 
 	cltd
-	divl %ebx
+	divl 112(%esp)
 	decl %ecx
-	movb (%edx,%ebp),%dl	# bh is always zero as ebx=radix < 36 
+	movb _dig_vec@GOTOFF(%ebx,%edx),%dl
 	testl %eax,%eax
 	movb %dl,(%ecx)
 	jne .L154
@@ -221,3 +230,13 @@
 
 .L10end:
 	.size	 longlong10_to_str,.L10end-longlong10_to_str
+
+	.section .gnu.linkonce.t.__i686.get_pc_thunk.bx,"ax",@progbits
+.globl __i686.get_pc_thunk.bx
+	.hidden	__i686.get_pc_thunk.bx
+	.type	__i686.get_pc_thunk.bx, @function
+__i686.get_pc_thunk.bx:
+	movl	(%esp), %ebx
+	ret
+
+	.section	.note.GNU-stack,"",@progbits
diff -NuarwbB mysql.orig/strings/strings-x86.s mysql/strings/strings-x86.s
--- mysql.orig/strings/strings-x86.s	2005-05-13 04:32:40.000000000 -0700
+++ mysql/strings/strings-x86.s	2005-05-17 00:11:56.000000000 -0700
@@ -415,3 +415,5 @@
 	ret
 .strxmov_end:
 	.size	 strxmov,.strxmov_end-strxmov
+
+	.section	.note.GNU-stack,"",@progbits
